packages/integrations/node/src/http-server.ts

```diff
@@ -23,12 +25,55 @@ function parsePathname(pathname: string, host: string | undefined, port: number)
 
 export function createServer(
 	{ client, port, host, removeBase }: CreateServerOptions,
-	handler: http.RequestListener
+	handler: http.RequestListener,
+	trailingSlash: AstroUserConfig['trailingSlash']
 ) {
 	const listener: http.RequestListener = (req, res) => {
 		if (req.url) {
-			let pathname: string | undefined = removeBase(req.url);
-			pathname = pathname[0] === '/' ? pathname : '/' + pathname;
+			const [urlPath, urlQuery] = req.url.split('?');
+			const filePath = path.join(fileURLToPath(client), removeBase(urlPath));
+
+			let pathname: string;
+			let isDirectory = false;
+			try {
+				isDirectory = fs.lstatSync(filePath).isDirectory();
+			}
+			catch (err) { }
+
+			if (!trailingSlash) // should never happen
+				trailingSlash = 'ignore';
+
+			const hasSlash = urlPath.endsWith('/');
+			switch (trailingSlash) {
+				case "never":
+					if (isDirectory && hasSlash) {
+						pathname = urlPath.slice(0, -1) + (urlQuery ? "?" + urlQuery : "");
+						res.statusCode = 301;
+						res.setHeader('Location', pathname);
+					} else pathname = urlPath;
+					// intentionally fall through
+				case "ignore":
+					{
+						if (isDirectory && !hasSlash) {
+							pathname = urlPath + "/index.html";
+						} else
+							pathname = urlPath;
+					}
+					break;
+				case "always":
+					if (!hasSlash) {
+						pathname = urlPath + '/' +(urlQuery ? "?" + urlQuery : "");
+						res.statusCode = 301;
+						res.setHeader('Location', pathname);
+					} else
+						pathname = urlPath;
+				break;
+			}
+			pathname = removeBase(pathname);
+
+			if (urlQuery && !pathname.includes('?')) {
+				pathname = pathname + '?' + urlQuery;
+			}
```

Comment:
Can you elaborate on this line?

If we add the query right before `send` handles the pathname, wouldn't it prevent `send` from matching files?

Reply:
@msxdan I would be ready to approve after clarification on this

Reply:
After my rebase, it was failing the tests with it.

Reply:
That's a good point. I hadn't thought of it, but that's behavior I would want as well.

In the code, I see that the query data is added to the path right before performing a redirect. I see that you had written tests for it as well.

So, it looks good to me if it looks good to you.

Reply:
Just checked, looks good to me as well!